### PR TITLE
Iteration4

### PR DESCRIPTION
**Pull request by abounk**
Apply APScheduler.

## PR Type

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Others. (Please describe)

### Changes

1. Add scheduler.
2. All jobs can be view at Django admin page.
Screen Shot 2564-11-15 at 17 13 04
3. Shows logging when job is added.
4. reminderTime for Tag and Assignment can accept only 3 values.
5. When Assignment doesn't have reminderTime, it will use reminderTime from Tag instead.

### CODE REVIEW

**Review by SoSorryTT — APPROVED** [review]

File: garnbarn_api/tests/test_scheduler.py [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+
+
+def create_assignment(name, due_date=None, reminder=None, tag=None):
+    """Create a assignment"""
+    assignment = Assignment(assignment_name=name,
+                            due_date=due_date,
+                            reminder_time=reminder,
+                            tag=tag
+                            )
+    return assignment
```
Comment (SoSorryTT) [inline garnbarn_api/tests/test_scheduler.py:17]: Change docstring to """Create an assignment."""

File: garnbarn_api/tests/test_scheduler.py [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+
+
+class TestScheduler(TestCase):
+    def setUp(self):
+        """Create tag and assignments"""
+        self.due_date = timezone.now() + timedelta(days=10)
+        hour = 3600  # seconds
+        day = 86400  # seconds
+        week = 604800  # seconds
```
Comment (SoSorryTT) [inline garnbarn_api/tests/test_scheduler.py:26]: Change docstring to """Create setup tag and assignments."""

File: garnbarn_api/tests/test_scheduler.py [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+                                              )
+
+    def datetime_before_due_date(self, duration):
+        """Get datetime before assignment's due date"""
+        if duration == "week":
+            return self.due_date - timedelta(days=7)
+        elif duration == "day":
+            return self.due_date - timedelta(days=1)
+        elif duration == "hour":
+            return self.due_date - timedelta(hours=1)
```
Comment (SoSorryTT) [inline garnbarn_api/tests/test_scheduler.py:54]: Change docstring to """Get datetime before assignment's due date."""

File: garnbarn_api/tests/test_scheduler.py [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+
+    def test_schedule_assignment_with_due_date_and_reminder_time(self):
+        """Schedule assignment with specified due date and reminder time"""
+        self.assignment_1.save()
+        jobs = scheduler.get_jobs()
+        self.assertEqual(3, len(jobs))
+        self.assertEqual(self.week_before_due_date, jobs[0].next_run_time)
+        self.assertEqual(self.hour_before_due_date, jobs[1].next_run_time)
+        self.assertEqual(self.assignment_1.due_date, jobs[2].next_run_time)
```
Comment (SoSorryTT) [inline garnbarn_api/tests/test_scheduler.py:63]: Change docstring to """Schedule assignment with specified due date and reminder time."""

File: garnbarn_api/tests/test_scheduler.py [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+
+    def test_schedule_assignment_with_no_due_date(self):
+        """Schedule must not be added if due date is not specified"""
+        self.assignment_2.save()
+        self.assertEqual([], scheduler.get_jobs())
```
Comment (SoSorryTT) [inline garnbarn_api/tests/test_scheduler.py:68]: Change docstring to """Schedule must not be added if due date is not specified."""

File: garnbarn_api/tests/test_scheduler.py [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+
+    def test_schedule_assignment_without_specifing_reminder_time(self):
+        """If reminder time of assignment is not provided, will use reminder time from tag instead."""
+        self.assignment_3.save()
+        jobs = scheduler.get_jobs()
+        self.assertEqual(2, len(jobs))
+        self.assertEqual(self.day_before_due_date, jobs[0].next_run_time)
+        self.assertEqual(self.assignment_3.due_date, jobs[1].next_run_time)
```
Comment (SoSorryTT) [inline garnbarn_api/tests/test_scheduler.py:76]: Change docstring to """If reminder time of assignment is not provided, assignment will use reminder time from tag instead."""

File: garnbarn_api/tests/test_scheduler.py [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+
+    def test_update_schedule(self):
+        """Update assignment reminder time will update the job schedules."""
+        self.assignment_1.save()
+        jobs = scheduler.get_jobs()
+        self.assertEqual(3, len(jobs))
```
Comment (SoSorryTT) [inline garnbarn_api/tests/test_scheduler.py:82]: Change docstring to """Update assignment reminder time will also update the job schedules."""

**Review by SoSorryTT — CHANGES_REQUESTED** [review]
Change docstring on `test_scheduler.py`

File: garnbarn_api/models.py [hunk 8]
```diff
@@ -101,3 +108,37 @@ def get_json_data(self):
...
+
+        for i in range(4):
+            job_id = f"Notification - {self.pk}_{i}"
+            if scheduler.get_job(job_id) is None:
+                break
+            scheduler.remove_job(job_id)
+
+        for index, item in enumerate(schedule_date):
+            schedule = self.due_date.timestamp() - item
+            schedule = datetime.fromtimestamp(schedule)
```
Comment (sirateek) [inline garnbarn_api/models.py:139]: The front end said that the timestamp sent into the reminder time field is milliseconds before the due date but the `due_date.timestamp` will return the timestamp in second unit.

File: garnbarn_api/models.py [hunk 9]
```diff
@@ -101,3 +108,37 @@ def get_json_data(self):
...
+            if schedule > datetime.now():
+                job = scheduler.add_job(pubsub, trigger=DateTrigger(run_date=schedule), id=f"Notification - {self.pk}_{index}",
+                                        max_instances=1, replace_existing=True)
+                logger.info(
+                    f"Schedule for assignment with id:{self.pk} has been set to trigger on ({job.next_run_time})")
```
Comment (sirateek) [inline garnbarn_api/models.py:144]: Maybe Add log said that which job is skipped (`DEBUG` level).

Comment (sirateek) [reply garnbarn_api/models.py:144]: Also, Reversing the logic and Moving the code up 1 level is also recommended

Comment (abounk) [reply garnbarn_api/models.py]: sloved 😎

File: garnbarn_api/models.py (hunk 8 above)
Comment (abounk) [reply garnbarn_api/models.py:139]: @TopsonArcana Fixed this in frontend.

**Review by sirateek — APPROVED** [review]
LGTM

**Review by SoSorryTT — APPROVED** [review]
Look good to me.

**Pull request merged**